demos/fleet_provisioning/fleet_provisioning_csr/fleet_provisioning_csr_demo.c

```diff
@@ -0,0 +1,1248 @@
+
+static void provisioningPublishCallback( MQTTPublishInfo_t * pPublishInfo,
+                                         uint16_t packetIdentifier )
+{
+    FleetProvisioningStatus_t status;
+    FleetProvisioningTopic_t api;
+
+    /* Silence compiler warnings about unused variables. */
+    ( void ) packetIdentifier;
+
+    status = FleetProvisioning_MatchTopic( pPublishInfo->pTopicName,
+                                           pPublishInfo->topicNameLength, &api );
+
+    if( status != FleetProvisioningSuccess )
+    {
+        LogError( ( "Unexpected publish message received. Topic: %.*s.",
+                    ( int ) pPublishInfo->topicNameLength,
+                    ( const char * ) pPublishInfo->pTopicName ) );
+    }
+    else
+    {
+        if( api == FleetProvCborCreateCertFromCsrAccepted )
+        {
+            LogInfo( ( "Fleet Provisioning CreateCertificateFromCsr accepted." ) );
+
+            responseStatus = ResponseAccepted;
+
+            ( void ) memcpy( ( void * ) payloadBuffer,
+                             ( const void * ) pPublishInfo->pPayload,
+                             ( size_t ) pPublishInfo->payloadLength );
+
+            payloadLength = pPublishInfo->payloadLength;
+        }
+        else if( api == FleetProvCborCreateCertFromCsrRejected )
+        {
+            LogError( ( "Fleet Provisioning CreateCertificateFromCsr rejected." ) );
+
+            responseStatus = ResponseRejected;
+        }
+        else if( api == FleetProvCborRegisterThingAccepted )
+        {
+            LogInfo( ( "Fleet Provisioning RegisterThing accepted." ) );
+
+            responseStatus = ResponseAccepted;
+
+            ( void ) memcpy( ( void * ) payloadBuffer,
+                             ( const void * ) pPublishInfo->pPayload,
+                             ( size_t ) pPublishInfo->payloadLength );
+
+            payloadLength = pPublishInfo->payloadLength;
+        }
+        else if( api == FleetProvCborCreateCertFromCsrRejected )
+        {
+            LogError( ( "Fleet Provisioning RegisterThing rejected." ) );
+
+            responseStatus = ResponseRejected;
+        }
+        else
+        {
+            LogError( ( "Unexpected Fleet Provisioning message. Topic: %.*s.",
+                        ( int ) pPublishInfo->topicNameLength,
+                        ( const char * ) pPublishInfo->pTopicName ) );
+        }
+    }
+}
+/*-----------------------------------------------------------*/
+
+static bool getCsr( char * buffer,
+                    size_t bufferLength,
+                    size_t * outCsrLength )
+{
+    FILE * file;
+    size_t length = 0;
+    bool status = true;
+
+    file = fopen( PROVISIONING_CSR_PATH, "rb" );
+
+    if( file == NULL )
+    {
+        LogError( ( "Error opening file at PROVISIONING_CSR_PATH." ) );
+        status = false;
+    }
+    else
+    {
+        int result;
+        result = fseek( file, 0L, SEEK_END );
```

Comment:
Comments at each of the POSIX file I/O calls will be helpful about their purpose